### ISSUE TITLE
PEP 544: Assignment to instance or class variables within a Protocol class if previously declared by a variable annotation

### ISSUE THREAD

**Issue by paul-dest** [microsoft/pyright#1618]
**Description of the alleged bug**
Regarding Protocol Members, [PEP 544](https://www.python.org/dev/peps/pep-0544/#protocol-members) states:

> To define a protocol variable, one can use PEP 526 variable annotations in the class body. Additional attributes _only_ defined in the body of a method by assignment via self are not allowed. The rationale for this is that the protocol class implementation is often not shared by subtypes, so the interface should not depend on the default implementation.

According to that, attributes previously declared by a variable annotation are _not_ disallowed to be assigned to within a Protocol class.

In contrast to that, pyright reports in all cases of an assignment:
`error: Assignment to instance or class variables not allowed within a Protocol class`

**To Reproduce**

```python
from typing import Protocol


class Base(Protocol):
    s: str

    def __init__(self, s: str) -> None:
        self.s = s

```
```
$ pyright test.py
No configuration file found.
stubPath /home/p---s/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/home/p---s/test.py
  8:14 - error: Assignment to instance or class variables not allowed within a Protocol class
1 error, 0 warnings, 0 infos 
Completed in 0.64sec
```

**Expected behavior**
no errors are reported from pyright

**Pyright version**
1.1.117

**Comment by erictraut** [comment]
Thanks for the bug report. I misinterpreted that line in PEP 544 when I implemented Protocol support in pyright. In particular I missed the key word `only`. In rereading the PEP, I agree with your interpretation. This will be fixed in the next release of pyright.

**Comment by paul-dest** [comment]
Meanwhile, I would like to suppress all `Assignment to instance or class variables not allowed within a Protocol class` errors. Unfortunately, I could not find an appropriate option in [Pyright Configuration](https://github.com/microsoft/pyright/blob/master/docs/configuration.md). Did I overlook the option?

**Comment by erictraut** [comment]
That diagnostic check is covered under the general rule "reportGeneralTypeIssues", so you would need to silence that entire class of errors.

I release a new version of pyright a couple of times every week, so you won't need to wait long before this is fixed.

**Comment by paul-dest** [comment]
That's very agile, thanks, @erictraut !

I have another test case for the bugfix:
```python
class MatcherProtocol(Protocol):
    qmin: int
    qmax: int

    def __new__(cls, *pargs, **kwargs):
        obj = super().__new__(cls)
        obj.qmin = kwargs.get('qmin', 1)
        obj.qmax = kwargs.get('qmax', 1)
        assert obj.qmin <= obj.qmax
        return obj
```

Currently, pyright reports the error mentioned above for each of the lines `obj.qmin = ...` and `obj.qmax = ...`.
Expected behavior is: no errors are reported from pyright.

I do not necessarily expect pyright to recognize, that `obj` is an instance complying to the `MatcherProtocol` (except maybe after an explicit annotation `obj: 'MatcherProtocol' = super().__new__(cls)`), but at least an assignment to the attributes of `obj` should not cause pyright to report errors.

**Comment by erictraut** [comment]
Thanks for the additional test case. I can confirm that my recent fix works as you have indicated. No errors are generated for this code sample.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.121, which I just published. It will also be included in the next release of pylance.
